### PR TITLE
Include duration in the ramping constraints

### LINKED ISSUE #772: Include durations on the ramping constraints

### PR DESCRIPTION
**Pull request by datejada**
# Pull request details

## Describe the changes made in this pull request

Include the duration in the ramping constraints in the RHS of the equations to account for that the ramp parameters are given as rates.

## List of related issues or pull requests

Closes #772 

## Collaboration confirmation

As a contributor I confirm

-   [x] I read and followed the instructions in README.dev.md
-   [x] The documentation is up to date with the changes introduced in this Pull Request (or NA)
-   [x] Tests are passing
-   [x] Lint is passing

### CODE REVIEW

File: src/create-model.jl [hunk 1]
```diff
@@ -266,6 +267,11 @@ function add_expression_terms_intra_rp_constraints!(
 
     for case in cases
         df_cons[!, case.col_name] .= JuMP.AffExpr(0.0)
+        conditions_to_add_min_outgoing_flow_duration =
+            add_min_outgoing_flow_duration && case.col_name == :outgoing_flow
+        if conditions_to_add_min_outgoing_flow_duration
+            df_cons[!, :min_outgoing_flow_duration] .= 1
```
Comment (abelsiqueira) [inline src/create-model.jl:273]:
```suggestion
            df_cons[!, :min_outgoing_flow_duration] .= 1.0
```

File: src/create-model.jl [hunk 2]
```diff
@@ -276,6 +282,9 @@ function add_expression_terms_intra_rp_constraints!(
             for i in eachindex(workspace)
                 workspace[i] = JuMP.AffExpr(0.0)
             end
+            if conditions_to_add_min_outgoing_flow_duration
+                outgoing_flow_durations = Vector()
+            end
```
Comment (abelsiqueira) [inline src/create-model.jl:287]:
```suggestion
            outgoing_flow_durations = Int
```
Or a large number

File: src/create-model.jl (hunk 1 above)
Comment (abelsiqueira) [reply src/create-model.jl:273]: Nevermind, it's supposed to be integer, right?

File: src/create-model.jl [hunk 3]
```diff
@@ -298,11 +307,17 @@ function add_expression_terms_intra_rp_constraints!(
                         row.flow,
                         resolution * efficiency_coefficient,
                     )
+                    if conditions_to_add_min_outgoing_flow_duration
+                        push!(outgoing_flow_durations, length(row.timesteps_block))
+                    end
```
Comment (abelsiqueira) [inline src/create-model.jl:312]:
```suggestion
                    if conditions_to_add_min_outgoing_flow_duration
                        outgoing_flow_durations = min(outgoing_flow_durations, length(row.timesteps_block))
                    end
```

File: src/create-model.jl [hunk 4]
```diff
@@ -298,11 +307,17 @@ function add_expression_terms_intra_rp_constraints!(
...
                 end
             end
             # Sum the corresponding flows from the workspace
             for row in eachrow(sub_df)
                 row[case.col_name] = agg(@view workspace[row.timesteps_block])
+                if conditions_to_add_min_outgoing_flow_duration
+                    row[:min_outgoing_flow_duration] = minimum(outgoing_flow_durations)
```
Comment (abelsiqueira) [inline src/create-model.jl:319]:
```suggestion
                    row[:min_outgoing_flow_duration] = outgoing_flow_durations
```

**Review by abelsiqueira — COMMENTED** [review]
Summary of the comments:

- Instead of storing a whole vector and computing the minimum, start with a large value and update the minimum every time. It's faster than `push!` and better for memory. An alternative would be to store a whole vector with predetermined size, but I am not sure that would be faster, and it would definitely use more memory.

File: src/create-model.jl (hunk 1 above)
Comment (datejada) [reply src/create-model.jl:273]: correct, it is integer 😄

Comment (abelsiqueira) [reply src/create-model.jl:273]: 👍

**Comment by datejada** [comment]
> Summary of the comments:
> 
> * Instead of storing a whole vector and computing the minimum, start with a large value and update the minimum every time. It's faster than `push!` and better for memory. An alternative would be to store a whole vector with predetermined size, but I am not sure that would be faster, and it would definitely use more memory.

@abelsiqueira, thanks for the suggestion. The PR is updated and rebased to the latest version in main 😄 Please let me know if you have more comments.

**Review by abelsiqueira — APPROVED** [review]
Thanks for the PR, LGTM.

**Pull request merged**